hermes/probe/monitor.go

```diff
@@ -12,9 +12,155 @@
+		return fmt.Errorf("invalid argument: opts.ProbeConf is not of type *probepb.HermesProbeDef")
+	}
+	mp.name = name
+	mp.config = conf
+	mp.targets = mp.config.GetTargets()
+	mp.opts = opts
+	mp.probeOpLatency = make(map[string]map[string][]*metrics.EventMetrics)
+	mp.apiCallLatency = make(map[string]map[string][]*metrics.EventMetrics)
+	mp.logger = opts.Logger
+
+	return nil
+}
+
+// Start runs the probe indefinitely, unless cancelled, at the configured interval.
+// Probe metrics will be sent via the metricChan at the beginning of the next probe run.
+// Arguments:
+//	- ctx: context provided for cancelling probe.
+//	- metricChan: bidirectional channel used for sending metrics to be surfaced.
+func (mp *MonitorProbe) Start(ctx context.Context, metricChan chan *metrics.EventMetrics) {
+	probeTicker := time.NewTicker(mp.opts.Interval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			probeTicker.Stop()
+			return
+		case <-probeTicker.C:
+			for _, target := range mp.probeOpLatency {
+				for _, op := range target {
+					for _, metric := range op {
+						metricChan <- metric
+					}
+				}
+			}
+
+			for _, target := range mp.apiCallLatency {
+				for _, call := range target {
+					for _, metric := range call {
+						metricChan <- metric
+					}
+				}
+			}
+
+			probeCtx, _ := context.WithDeadline(ctx, time.Now().Add(mp.opts.Timeout))
+			// init during probe run as metrics are timestamped.
+			if err := mp.initProbeMetrics(); err != nil {
+				mp.logger.Errorf("error initialising probe metric: %v", err)
+			}
+			mp.runProbe(probeCtx)
+		}
+	}
+}
+
+// initLatencyDistMetric initialises a latency distribution metric
+// according to the values passed.
+// Arguments:
+//	- target: pass target name.
+//	- latencyMap: pass a map with the key format: map[target][operation].
+//	- latDistPb: pass the metric proto distribution to be read from.
+//	- operations: pass the list of operations.
+// Returns:
+//	- error: returns an error if the proto distribution variable passed is invalid.
+func (mp *MonitorProbe) initLatencyDistMetric(target string, latencyMap map[string]map[string][]*metrics.EventMetrics, latDistPb *metricpb.Dist, operations []string) error {
+	if latencyMap[target] != nil {
+		return nil
+	}
+
+	latDist, err := metrics.NewDistributionFromProto(latDistPb)
+	if err != nil {
+		return fmt.Errorf("invalid argument: error creating probe latency distribution from the specification (%v): %v", latDistPb, err)
+	}
+	for _, op := range operations {
+		latencyMap[target][op] = append(latencyMap[target][op], metrics.NewEventMetrics(time.Now()).
+			AddMetric("hermes_probe_operation_latency_s", latDist.Clone()).
+			AddLabel("storage_system", mp.config.GetTargetSystem().String()).
+			AddLabel("probe_operation_type", op).
+			AddLabel("target", target))
+	}
+	return nil
+}
+
+// initProbeMetrics initializes missing latency metrics for each target.
+func (mp *MonitorProbe) initProbeMetrics() error {
+	for _, target := range mp.targets {
```

Comment:
Nit: Use short variable names for small-scoped variables. Just `t` would do here instead of `target`.

Reply:
Okay, thank you. I'll amend this.